### PR TITLE
Separate application from infrastructure

### PR DESCRIPTION
**Pull request by sjmiller609**
Closes: https://github.com/astronomer/issues/issues/1284

TODO:
- [ ] remove sjmiller609 image (by adding image build at start, including the stuff in bin/install)
- [ ] add lint step
- [ ] add push to dev cluster the configuration
- [x] add secrets to circle CI

Reviewing:
- read README.md first off, then review the rest

What does this do?
- add idempotent helm3 migration
- migrate from drone to circle CI
- separate application from infrastructure release
- set up secrets management for getting secrets into helm values from GCS bucket in a repeatable manner

### CODE REVIEW

File: bin/config.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        replacing the !Secret tags with the value of that secret, downloaded
+        from Google Cloud Storage. So, our 'representer' (yaml dumper)
+        uses the representer of strings, so that load then dump swaps out
+        the !Secret tag for the value stored in self.secret_value. See the
+        constructor for details on how this value is found.
+        '''
+        return dumper.represent_str(data.secret_value)
+
+yaml.SafeLoader.add_constructor(Secret.yaml_tag, Secret.from_yaml)
+yaml.SafeDumper.add_multi_representer(Secret, Secret.to_yaml)
```
Comment (sjmiller609) [inline bin/config.py:62]: this part is replacing the 'secrets.tf' functionality, previously used to add secrets into helm values. This should be easier for developers to work with and it's more DRY now with the shared configuration in a single file, and each environment-specific config having only the stuff specific to that environment. Secrets must live in the environment-specific configs.

File: bin/helm_cloud.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+DRY_RUN_FLAG=""
+if ! [ -z "$DRY_RUN" ]; then
+  DRY_RUN_FLAG="--dry-run"
+fi
+
+if ! [ -z "$RELEASES_TO_UPGRADE" ]; then
+  echo "Non zero Airflow and Astronomer releases on Helm 2. Performing Helm 2 to Helm 3 upgrade procedure"
+  echo "Scaling down ingress so nobody can access Astronomer, this avoids race conditions of the upgrade process against customer activity"
+  kubectl scale --replicas=0 -n astronomer deployment/astronomer-nginx
+  echo "Scaling down commander, this ensures that old commander can't be used during the upgrade procedure"
```
Comment (jpweber) [inline bin/helm_cloud.sh:40]: After the initial upgrade (which I think just happened) this stuff doesn't need to be here right?

File: .circleci/config.yml [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    environment:
+      BASE_DOMAIN: gcp0001.us-east4.astronomer.io
+      DEPLOYMENT_ID: prod
+      PROJECT: astronomer-cloud-prod
+    steps:
+      - checkout
+      - helm-upgrade
+  helm_dry_run_prod:
+    docker:
+      - image: sjmiller609/helm-kubectl-terraform:latest
```
Comment (schnie) [inline .circleci/config.yml:36]: Should we move these images into our dockerhub?

**Comment by schnie** [comment]
So if we need to add new secrets to helm values, whats the process for that and should we document that somewhere?

File: README.md [hunk 4]
```diff
@@ -1,3 +1,50 @@
...
+
+Secrets should not be committed to source code. If you want to add a secret into the Helm values, add a '!Secret' tag as the value, for example:
+
+```
+this:
+  is:
+    my: !Secret name_of_secret
+```
+
+Then, provide the name_of_secret and the value you want to Astronomer cloud infrastructure team. Astronomer cloud infrastructure team will upload an object named "name_of_secret.txt" to the Google Cloud Storage bucket. Automation will replace "!Secret" tags with the corresponding values, which are securely downloaded from the secrets bucket.
```
Comment (sjmiller609) [inline README.md:33]: @schnie I think this addresses your question

File: .circleci/config.yml (hunk 3 above)
Comment (sjmiller609) [reply .circleci/config.yml:36]: yes

File: bin/helm_cloud.sh (hunk 2 above)
Comment (sjmiller609) [reply bin/helm_cloud.sh:40]: yes agreed, this was only required on the first time

Comment (sjmiller609) [reply bin/helm_cloud.sh:40]: the $RELEASES_TO_UPGRADE var will be empty in subsequent runs. I'm leaving it in on account of sql proxy, istio, etc still running helm 2

**Pull request merged**